### PR TITLE
perf(sql): speed up negative limit queries with a wildcard

### PR DESCRIPTION
**Pull request by nwoolmer**
Fixes https://github.com/questdb/questdb/issues/4696

Raises this issue: https://github.com/questdb/questdb/issues/5138

`rewriteNegativeLimit` handled basic cases such as this:

```sql
EXPLAIN SELECT timestamp FROM trades LIMIT -3;
```

```
Radix sort light
  keys: [timestamp]
    Limit lo: 3
        PageFrame
            Row backward scan
            Frame backward scan on: trades
```

In this case, we first perform a reverse scan, then sort in the ascending direction.

However, if you add a wildcard:

```sql
EXPLAIN SELECT timestamp, * FROM trades LIMIT -3;
```

We get an unoptimised plan:

```
Limit lo: -3
    SelectedRecord
        PageFrame
            Row forward scan
            Frame forward scan on: trades
```

After these changes, we can still have an optimised plan:

```
Radix sort light
  keys: [timestamp]
    SelectedRecord
        Limit lo: 3
            PageFrame
                Row backward scan
                Frame backward scan on: trades
```

Adding a wildcard causes the model to look something like this:

```
select-choose timestamp, symbol, side, price, amount, timestamp timestamp1 
from (select-choose timestamp, symbol, side, price, amount 
from (trades timestamp (timestamp))) limit -(3)
```

Unfortunately, this isn't caught by earlier checks in `rewriteSelectClause` around merging select-choose and group by, since its not a group by But one of these models is unnecessary. The changes in this PR add the possibility of condensing the models together.

### CODE REVIEW

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 58 / 71 (81.69%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|58|71|81.69%|

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 1]
```diff
@@ -5736,6 +5792,93 @@ private void rewriteSingleFirstLastGroupBy(QueryModel model) {
         }
     }
 
+    private void rewriteToCondenseWildcardModels(QueryModel model) throws SqlException {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:5795]: Would be great to add some javadoc for this method.

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 2]
```diff
@@ -5736,6 +5792,93 @@ private void rewriteSingleFirstLastGroupBy(QueryModel model) {
...
+                                    copiedColumns.add(nested.getBottomUpColumns().size() - 1);
+                                }
+                            }
+                        }
+                    }
+
+                    boolean identicalModels = true;
+
+                    // now check that the models are identical
+                    if (model.getBottomUpColumns().size() == nested.getBottomUpColumns().size()) {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:5854]: What happens when `model.getBottomUpColumns().size() != nested.getBottomUpColumns().size()`? We end up with `identicalModels` set to true, so we get rid of the nested model. Is that the desired behavior?

File: core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java [hunk 3]
```diff
@@ -2572,6 +2572,52 @@ public void testRangeFillWhenThereAreNoRecords() throws Exception {
...
+        assertMemoryLeak(() -> {
+            ddl("CREATE TABLE 'trades' (\n" +
+                    "  symbol SYMBOL,\n" +
+                    "  side SYMBOL,\n" +
+                    "  price DOUBLE,\n" +
+                    "  amount DOUBLE,\n" +
+                    "  timestamp TIMESTAMP\n" +
+                    ") timestamp (timestamp) PARTITION BY DAY WAL;");
+            drainWalQueue();
+            assertPlanNoLeakCheck("select timestamp, * from trades limit -3", "Radix sort light\n" +
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java:2611]: Would be great to add more cases, e.g. different order `select *, timestamp` or aliases `select timestamp ts1, timestamp ts2, *` and so on.

File: core/src/main/java/io/questdb/griffin/ExpressionParser.java [hunk 4]
```diff
@@ -1319,7 +1319,7 @@ This is only a problem for ambiguous operators like minus('-'), complement('~')
                             }
 
                             // If the token is a function token, then push it onto the stack.
-                            opStack.push(expressionNodePool.next().of(ExpressionNode.LITERAL, GenericLexer.immutableOf(tok), Integer.MIN_VALUE, lastPos));
+                            opStack.push(expressionNodePool.next().of(ExpressionNode.LITERAL, GenericLexer.immutableOf(tok), 0, lastPos));
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/ExpressionParser.java]: Yes, precedence is relevant for operators. For other nodes it shouldn't be relevant.

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 58 / 71 (81.69%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|58|71|81.69%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 67 (92.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|62|67|92.54%|

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 5]
```diff
@@ -5736,6 +5792,100 @@ private void rewriteSingleFirstLastGroupBy(QueryModel model) {
...
+                    && none.getGroupBy().size() == 0
+                    && model.getSampleBy() == null
+                    && nested.getSampleBy() == null
+                    && !hasAggregateQueryColumn(model)
+                    && !model.isDistinct()) {
+
+                IntList copiedColumns = new IntList();
+
+                // we are going to compact the models
+                if (model.getBottomUpColumns().size() >= nested.getBottomUpColumns().size()) {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:5848]: Can we move this condition to the large if block on L5830?

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 67 (92.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|62|67|92.54%|

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**